### PR TITLE
fix(core): correctly project single-root content inside control flow

### LINKED ISSUE #54840: Conditionals and content projection
### Which @angular/* package(s) are the source of the bug?

core

### Is this a regression?

No

### Description

There is a surprising interaction between our static (creation time) content projection system _and_ the new control flow (although this is _not_ specific to the new control flow - the same can be observed with `ngIf`).

To illustrate the surprising scenario let's consider a component with content projection with a selector (`<ng-content select="some-content">`) and content to project wrapped in the control flow `@if` statement with the `else` branch:

```html
<with-content>
      @if(false) {          
        <some-content />
      } @else {
        I'm not the content queried for, don't project me! 
      }
</with-content>
```

https://stackblitz.com/edit/stackblitz-starters-57dynz?file=src%2Fmain.ts

As surprising as it may be, the ` I'm not the content queried for, don't project me!` text node will be taken as content to project (!). This is consistent with the existing implementation which will project the whole `@if` container and container capturing the `some-content` as a tag name of a root node.

While we can't fix _all_ issues with the content projection, we could remedy this one by 

### Please provide a link to a minimal reproduction of the bug

https://stackblitz.com/edit/stackblitz-starters-57dynz?file=src%2Fmain.ts

### Please provide the exception or error you saw

_No response_

### Please provide the environment you discovered this bug in (run `ng version`)

_No response_

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by pkozlowski-opensource**
This commit changes the way we use containers to insert conditional content. Previously if and switch conditional would always use the first content as the insertion container. This strategy interfered with content projection that projects entire containers - as the consequence content for _all_ cases would end up in slot matched by the first container. This could be very surprising as described in https://github.com/angular/angular/issues/54840

After the change each conditional content is projected into its own container. This means that content projection can match more than one container and result in correct display.

Fixes #54840

### CODE REVIEW

**Comment by pkozlowski-opensource** [comment]
@crisbeto this PR has runtime bits + tests of the changed strategy of inserting conditional content into containers. I've added tests for `@if` and `@switch` - `@for` should be fine as far as I can tell.

Important notes:
* DO NOT MERGE this PR on its own - it requires compiler changes as well;
* with those changes we could simplify the `ɵɵconditional` instruction by dropping its first `containerIndex` argument

File: packages/core/src/render3/instructions/control_flow.ts [hunk 1]
```diff
@@ -32,48 +33,55 @@ import {ɵɵtemplate} from './template';
...
  * removing views selected by a conditional expression.
  *
- * @param containerIndex index of a container in a host view (indexed from HEADER_OFFSET) where
- *     conditional views should be inserted.
  * @param matchingTemplateIndex index of a template TNode representing a conditional view to be
  *     inserted; -1 represents a special case when there is no view to insert.
  * @codeGenApi
  */
 export function ɵɵconditional<T>(containerIndex: number, matchingTemplateIndex: number, value?: T) {
+  // TODO: we could remove the containerIndex argument to this instruction now (!)
```
Comment (crisbeto) [inline packages/core/src/render3/instructions/control_flow.ts:41]: Nice! I would probably do it in a standalone PR though.

File: packages/core/test/acceptance/control_flow_if_spec.ts [hunk 2]
```diff
@@ -378,15 +376,12 @@ describe('control flow - if', () => {
       expect(fixture.nativeElement.textContent).toBe('Main: Before  After Slot: foo');
     });
 
-    // Note: the behavior in this test is *not* intuitive, but it's meant to capture
-    // the projection behavior from `*ngIf` with `@if`. The test can be updated if we
-    // change how content projection works in the future.
-    it('should project an @else content into the slot of @if', () => {
+    xit('should project @id an @else content into separate slots', () => {
```
Comment (crisbeto) [inline packages/core/test/acceptance/control_flow_if_spec.ts:379]: `@id` -> `@if`. I can also fix it when I cherry pick.

File: packages/core/src/render3/instructions/control_flow.ts (hunk 1 above)
Comment (pkozlowski-opensource) [reply packages/core/src/render3/instructions/control_flow.ts:41]: Yep, make sense to do it in a separate PR!

File: packages/core/test/acceptance/control_flow_if_spec.ts (hunk 2 above)
Comment (pkozlowski-opensource) [reply packages/core/test/acceptance/control_flow_if_spec.ts:379]: Ooops, fixed - thnx!

**Comment by crisbeto** [comment]
Compiler changes plus the changes from this PR are in https://github.com/angular/angular/pull/54921.

**Comment by pkozlowski-opensource** [comment]
Landed as part of #54921
